pennylane_cirq/simulator_device.py

```diff
@@ -71,225 +59,115 @@ class SimulatorDevice(CirqDevice):
-
-        for marginal_state in marginal_states:
-            marginal_probabilities[marginal_state] = 0.0
-
-        for state in probabilities:
-            marginal_state = tuple(state[wire] for wire in wires)
-            marginal_probabilities[marginal_state] += probabilities[state]
-
-        return marginal_probabilities
-
-    def _get_eigenvalues(self, observable, wires, par):
-        """Return the eigenvalues of the given observable.
+            self._state = np.array(self._result.state_vector())
 
-        Args:
-            observable (str or list[str]): name of the observable(s)
-            wires (List[int] or List[List[int]]): subsystems the observable(s) is to be measured on
-            par (tuple or list[tuple]]): parameters for the observable(s)
+    def probability(self, wires=None):
+        if self._state is None:
+            return None
 
-        Returns:
-            array[float]: eigenvalues of the observable
-        """
-        num_wires = len(wires)
+        wires = wires or range(self.num_wires)
+        probs = np.abs(self._state) ** 2
 
-        # All ones corresponds to the Identity observable
-        eigenvalues = np.ones(2 ** num_wires)
+        return self.marginal_prob(probs, wires)
 
-        if observable == "Hermitian":
-            # Take the eigenvalues from the stored values
-            Hmat = par[0]
-            Hkey = tuple(Hmat.flatten().tolist())
-
-            eigenvalues = self._eigs[Hkey]["eigval"]
-        elif observable != "Identity":
-            # If we don't have an Hermitian observable we use
-            # a diagonalization to tensors of Z observables
-            eigenvalues = z_eigs(num_wires)
-
-        return eigenvalues
-
-    def expval(self, observable, wires, par):
-        eigenvalues = self._get_eigenvalues(observable, wires, par)
-
-        if self.analytic:
-            # We have to use the state of the simulation to find the expectation value
-            marginal_probability = np.fromiter(
-                self.marginal_probability(wires).values(), dtype=np.float
-            )
-
-            return np.dot(eigenvalues, marginal_probability)
-        else:
-            return self.sample(observable, wires, par).mean()
-
-    def var(self, observable, wires, par):
-        eigenvalues = self._get_eigenvalues(observable, wires, par)
+    @property
+    def state(self):
+        return self._state
 
+    def generate_samples(self):
         if self.analytic:
-            # We have to use the state of the simulation to find the expectation value
-            marginal_probability = np.fromiter(
-                self.marginal_probability(wires).values(), dtype=np.float
-            )
-
-            return (
-                np.dot(eigenvalues ** 2, marginal_probability)
-                - np.dot(eigenvalues, marginal_probability) ** 2
-            )
-        else:
-            return self.sample(observable, wires, par).var()
+            return super().generate_samples()
```

Comment:
How about the case of `analytic==True` and returning `qml.sample`?

Reply:
I see no problem with that? The logic of deciding when to call `generate_samples` is implemented in `QubitDevice`, so I just have to assure that samples are properly generated when it is called. And for analytic mode I can safely piggyback onto the default implementation.

Reply:
This is the part of the `device.execute()`:
```python
        if (not self.analytic) or circuit.is_sampled:
            self._samples = self.generate_samples()
```
for `self.analytic==True and circuit.is_sampled` `SimulatorDevice.generate_samples()` will be called and since `self.analytic==True` the if statement evaluates to `True` on line 158, and `QubitDevice.generate_samples()` will be called.

Whereas, if `self.analytic==False and circuit.is_sampled`, `SimulatorDevice.generate_samples()` still gets called, but the if statement on line 158 evaluates to `False` and the code block below gets executed.

If this is so (and I didn't confuse myself or mess up something here :sweat_smile: ), then how come that there is a different logic for these two cases?

Reply:
All you say is totally correct. There is a different logic for the `analytic==False` case because in that case I don't generate the samples myself but I use Cirq's own measurement scheme. I did this because I wanted to use as much of Cirq's inbuilt API as possible

Reply:
This is the difference between `simulator.run` (`analytic==False`) and `simulator.simulate` (`analytic==True`)